### ISSUE TITLE
Slow commit

### ISSUE THREAD

**Issue by Artimi** [aio-libs/aiokafka#128]
Hi,
I was digging in our messaging system and I noticed that `consumer.commit()` takes quite a lot of time. In my informal benchmarking I saw that it takes about 100 ms but I also saw 500 ms delay. I tried to compare it with `kafka-python` on the same topic at the same time and it took about 1 ms. Do you have any idea why it can take so long? Does commit in `aiokafka` do more than commit in `kafka-python`? We want to call commit after every message and this is really damaging performance. I would like to confirm every message and not just use auto commit.

While researching this I discovered that Java client and also kafka-python has so called `ConsumerCoordinator.commit_offsets_async()` https://github.com/dpkp/kafka-python/blob/master/kafka/coordinator/consumer.py#L334 . As I understand it is fire-and-forget commit that does not wait for the response from kafka server. This might help our perfomance. Why it is not present in `aiokafka`?

**Comment by Drizzt1991** [comment]
@Artimi please provide at least some snippets to understand how you measured it. Maybe you did not wait for actual commit on `kafka-python`, or were in a group rejoin on `aiokafka`. I can only speculate at this point.
As for your question about why there is no `commit_offsets_async `, it's simple - cause asyncio is asynchronous by design, you can just run it commit in a task.

**Comment by Drizzt1991** [comment]
If you need additional info on the topic, here's a few tips:
* when Consumer commits info it's actually a publish to a special internal topic in Kafka, so it should be very fast. 
* Commit is executed using special API, that will send the request to a special Coordinator node. This means that if a Group Join is in process (it's executed via Coordinator too) you must wait for that to finish first.
* As the protocol for any connection to a broker in Kafka is synchronous it means, that if you have some pending operation going on to that node, you will have to wait for it to finish. This can lead to an idea, that you are waiting for data with a timeout_ms=500 on the same node you want to commit to. This can result in such behaviour.

Understand that `aiokafka` is an asynchronous library and we pipeline all operations instead of only sending/receiving on `consumer.poll()` call as `kafka-python` does it. `kafka-python` may just wait for previous operation to end where aiokafka will give control to you as soon as possible.

**Comment by Drizzt1991** [comment]
Hey there, any progress on this?

**Comment by Artimi** [comment]
Hey sorry about be silent. I've discovered that in somehow our enhanced client sometimes calls `FetchRequest` that waits up to 500 ms and so `OffsetCommitRequest` is blocked until `FetchRequest` is "timeouted". We probably do something wrong because it does not happen with pure aiokafka. Sorry about that. I hope I'll solve this and let you know what was our mistake. Thanks for response.

**Comment by Artimi** [comment]
Hey @Drizzt1991, so I've found precise cause of our problems. Normally reading from aiokafka when you want to commit everything is like this:
```python
while True:
    msg = await consumer.getone()
    process(msg)
    await consumer.commit()
```
This way `FetchRequest` is send always after `OffsetCommitRequest` and therefore `OffsetCommitRequest` is not blocked by `FetchRequest`. In our use case we need single consumer that listens to multiple topics that can change through time. To solve this we have coroutine that calls `getone` on all topics and puts them in internal queue. This means that `getone` and `commit` are not sequential in code but may occur concurrently. And it is a problem because when `OffsetCommitRequest` is send right after `FetchRequest` it has to wait 500 ms sometimes. To test this I tried to put `commit` in `ensure_future` to model our use case (although it is another way in our use case: we have while loop in coroutine and commit is called from "main thread"):
```python
async def commit_delayed(consumer):
	await asyncio.sleep(0.05)
	await consumer.commit()

while True:
    msg = await consumer.getone()
    process(msg)
   asyncio.ensure_future(commit_delayed(consumer))
```
As you said the protocol to Kafka broker is synchronous and therefore I don't think this can be somehow fixed.
About asynchronous commit I was asking: we didn't want to use task because when something like this happens (connection is blocked for some time) we would have created a lot of these tasks and it would take some to process those and it could slow down other tasks in asyncio loop. Right now we will probably solve this with debounced commit (task that gets called only after some time but if this task is called again meanwhile it is delayed up to some maximum time). 
Thanks for your suggestions I really appreciate your help. Sorry about that I blamed aiokafka for this. At least I have opportunity to learn the library really well.

**Comment by Drizzt1991** [comment]
@Artimi Thanks for the feedback. I'll let the bug stick around for some time, as this is an issue for a lot of usecases. It's probably possible to have separate sockets for **Fetching** and **Commiting** or at least a way to disable **Fetch** if you don't want it to block commits.
I'll try to look into it for possible workarounds.

**Comment by Drizzt1991** [comment]
Ok, I did a bit of digging in Java's Kafka client. Java's client really does separate Coordinator and Consuming sockets, so that coordinator is never blocked or hindered.
References:
https://issues.apache.org/jira/browse/KAFKA-1925
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java#L575

**Comment by tvoinarovskyi** [comment]
Should be fixed by #196.

**Comment by pkit** [comment]
Using the latest `aiokafka==0.12.0` I get 2.5 sec for 100 message commits.
I.e. on average 250ms per commit. All localhost on my laptop.
Seems pretty bad to me.
